src/compiler/transformer.ts

```diff
@@ -154,51 +168,94 @@ namespace ts {
             emit(node);
         }
 
-        /**
-         * Gets flags that control emit behavior of a node.
-         */
-        function getNodeEmitFlags(node: Node) {
-            while (node) {
-                const nodeId = node.id;
-                if (nodeId && nodeEmitFlags[nodeId] !== undefined) {
-                    return nodeEmitFlags[nodeId];
+        function getNodeEmitOptions(node: Node, createIfMissing: boolean) {
+            // Keeps track of the nearest set of options
+            let options: NodeEmitOptions;
+            let currentNode = node;
+            while (currentNode) {
+                const currentOptions = currentNode.emitOptions || nodeEmitOptions[getNodeId(currentNode)];
+                if (currentOptions) {
+                    options = currentOptions;
+                    break;
+                }
+
+                currentNode = currentNode.original;
+            }
+
+            if (currentNode !== node && createIfMissing) {
+                options = options ? clone(options) : { };
+                if (isSourceTreeNode(node)) {
+                    nodeEmitOptions[getNodeId(node)] = options;
+                }
+                else {
+                    node.emitOptions = options;
                 }
+            }
+
+            // Merge with previous options on get.
+            if (options && options.flags & NodeEmitFlags.Merge) {
+                const previousOptions = getNodeEmitOptions(currentNode.original, /*createIfMissing*/ false);
+                if (previousOptions) {
+                    options.flags = (options.flags | previousOptions.flags) & ~NodeEmitFlags.Merge;

```

Comment:
question: The reason why we negate NodeEmitFlags.Merge is so that we don't keep merge all NodeEmitOptions?